code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/cluster/ClusterAddHostActionHandler.java

```diff
@@ -0,0 +1,62 @@
+package io.cattle.platform.iaas.api.cluster;
+
+import java.util.Map;
+
+import javax.inject.Inject;
+
+import io.cattle.platform.api.action.ActionHandler;
+import io.cattle.platform.core.constants.ClusterConstants;
+import io.cattle.platform.core.dao.ClusterHostMapDao;
+import io.cattle.platform.core.model.ClusterHostMap;
+import io.cattle.platform.core.model.Host;
+import io.cattle.platform.object.ObjectManager;
+import io.github.ibuildthecloud.gdapi.exception.ValidationErrorException;
+import io.github.ibuildthecloud.gdapi.request.ApiRequest;
+import io.github.ibuildthecloud.gdapi.validation.ValidationErrorCodes;
+
+public class ClusterAddHostActionHandler implements ActionHandler {
+
+    ObjectManager objectManager;
+    ClusterHostMapDao clusterHostMapDao;
+
+    @Override
+    public String getName() {
+        return "host.addhost";
+    }
+
+    @Override
+    public Object perform(String name, Object obj, ApiRequest request) {
+        if (!objectManager.isKind(obj, ClusterConstants.KIND)) {
+            return null;
+        }
+        Host cluster = (Host)obj;
+
+        Long hostToAddId = (Long)((Map)request.getRequestObject()).get(ClusterConstants.ADD_REMOVE_HOST_PARAM);
+        Host hostToAdd = objectManager.loadResource(Host.class, hostToAddId);
+
+        // I'd prefer not to have this check and rely on the DB unique constraint below if possible
+        ClusterHostMap existingMapping = clusterHostMapDao.getClusterHostMap(cluster, hostToAdd);
+        if (existingMapping != null) {
+            throw new ValidationErrorException(ValidationErrorCodes.INVALID_REFERENCE, ClusterConstants.ADD_REMOVE_HOST_PARAM);
+        }
+
+        ClusterHostMap newMapping = objectManager.newRecord(ClusterHostMap.class);
+        newMapping.setClusterId(cluster.getId());
+        newMapping.setHostId(hostToAdd.getId());
+        newMapping.setState("created");
+
+        newMapping = objectManager.create(newMapping);
```

Comment:
In general we would like to still orchestrate the creation of mapping entities if possible.  Is there any downside of doing `io.cattle.platform.core.dao.GenericResourceDao.createAndSchedule(Class<T>, Map<String, Object>)`